File: apps/server/src/execution/flow/execution_planner.py
"""Execution planner integrating logic from dynamic_executor.py."""

from typing import List, Dict, Set, Optional, Any
from collections import defaultdict, deque
import logging

from ..core.execution_context import ExecutionContext
from .dependency_resolver import DependencyResolver
from apps.server.src.constants import NodeType
from apps.server.src.utils.arrow_utils import ArrowUtils

logger = logging.getLogger(__name__)


class ExecutionPlanner:
    """Plans execution order and handles execution planning logic."""
    
    def __init__(self, context: ExecutionContext, dependency_resolver: DependencyResolver):
        self.context = context
        self.dependency_resolver = dependency_resolver
    
    def create_execution_plan(self) -> Dict[str, Any]:
        """Create execution plan with validation and metadata (from dynamic_executor)."""
        start_nodes = self.dependency_resolver.validate_start_nodes()
        cycles = self.dependency_resolver.detect_cycles()

        return {
            "start_nodes": start_nodes,
            "has_cycles": len(cycles) > 0,
            "cycle_nodes": cycles,
            "total_nodes": len(self.context.nodes_by_id),
            "total_arrows": len(self._get_all_arrows()),
            "node_types": {node_id: node.get("type", "unknown") for node_id, node in self.context.nodes_by_id.items()}
        }
    
    def get_start_nodes(self) -> List[str]:
        """Get all start nodes in the diagram."""
        return [
            node_id
            for node_id, node in self.context.nodes_by_id.items()
            if node.get('type') == 'startNode'
        ]
    
    def get_next_nodes(self, node_id: str, output: Any = None) -> List[str]:
        """Get the next nodes to execute based on current node output."""
        next_nodes = []
        outgoing = self.context.outgoing_arrows.get(node_id, [])
        
        for arrow in outgoing:
            target_id = ArrowUtils.get_target(arrow)
            if not target_id:
                continue
                
            # Handle conditional arrows
            if self._is_conditional_arrow(arrow) and isinstance(output, bool):
                arrow_label = self._get_arrow_label(arrow)
                if (arrow_label == 'true' and output) or (arrow_label == 'false' and not output):
                    next_nodes.append(target_id)
            else:
                # Non-conditional arrow
                next_nodes.append(target_id)
        
        return next_nodes
    
    def _is_conditional_arrow(self, arrow: dict) -> bool:
        """Check if arrow is conditional based on label."""
        label = self._get_arrow_label(arrow)
        return label in ['true', 'false']
    
    def _get_arrow_label(self, arrow: dict) -> str:
        """Get arrow label for conditional logic."""
        # Check multiple possible locations for the label
        label = arrow.get('label', '')
        if not label:
            # Check sourceHandle for condition output patterns
            source_handle = arrow.get('sourceHandle', '')
            if source_handle.endswith('-output-true'):
                label = 'true'
            elif source_handle.endswith('-output-false'):
                label = 'false'
            else:
                # Check arrow data
                label = arrow.get('data', {}).get('label', '')
        
        return label.lower() if label else ''
    
    def _get_all_arrows(self) -> List[dict]:
        """Get all arrows from the context."""
        all_arrows = []
        for arrows in self.context.outgoing_arrows.values():
            all_arrows.extend(arrows)
        return all_arrows
    
    def detect_parallel_execution_opportunities(self) -> List[Set[str]]:
        """Identify nodes that can be executed in parallel."""
        parallel_groups = []
        processed = set()
        
        while len(processed) < len(self.context.nodes_by_id):
            # Find nodes with no unprocessed dependencies
            ready_nodes = set()
            
            for node_id in self.context.nodes_by_id:
                if node_id in processed:
                    continue
                
                # Check if all dependencies are processed
                deps = set(self.dependency_resolver.get_dependencies(node_id))
                if deps.issubset(processed):
                    ready_nodes.add(node_id)
            
            if not ready_nodes:
                # Handle cycles or disconnected nodes
                remaining = set(self.context.nodes_by_id.keys()) - processed
                if remaining:
                    ready_nodes.add(next(iter(remaining)))
            
            if ready_nodes:
                parallel_groups.append(ready_nodes)
                processed.update(ready_nodes)
        
        return parallel_groups
    
    def validate_execution_path(self, path: List[str]) -> Dict[str, Any]:
        """Validate if an execution path is valid."""
        validation_result = {
            "is_valid": True,
            "errors": [],
            "warnings": []
        }
        
        executed = set()
        
        for i, node_id in enumerate(path):
            if node_id not in self.context.nodes_by_id:
                validation_result["is_valid"] = False
                validation_result["errors"].append(f"Node {node_id} not found in diagram")
                continue
            
            # Check dependencies
            deps = set(self.dependency_resolver.get_dependencies(node_id))
            missing_deps = deps - executed
            
            if missing_deps:
                validation_result["is_valid"] = False
                validation_result["errors"].append(
                    f"Node {node_id} at position {i} has unmet dependencies: {missing_deps}"
                )
            
            executed.add(node_id)
        
        return validation_result

File: apps/web/src/engine/flow/execution-planner.ts
import { ExecutionContext, ExecutionPlan, DiagramArrow } from '@/shared/types/core';
import { DependencyResolver } from './dependency-resolver';

export class ExecutionPlanner {
  constructor(
    private context: ExecutionContext,
    private dependencyResolver: DependencyResolver
  ) {}

  /**
   * Create execution plan with validation and metadata
   */
  createExecutionPlan(): ExecutionPlan {
    const startNodes = this.dependencyResolver.validateStartNodes();
    const parallelGroups = this.detectParallelExecutionOpportunities();
    const executionOrder = this.determineExecutionOrder();
    
    // Create dependencies map
    const dependencies: Record<string, string[]> = {};
    for (const nodeId of Object.keys(this.context.nodesById)) {
      dependencies[nodeId] = this.dependencyResolver.getDependencies(nodeId);
    }

    return {
      executionOrder,
      parallelGroups: parallelGroups.map(group => Array.from(group)),
      dependencies,
      estimatedCost: 0, // TODO: Calculate based on node types
      estimatedTime: 0  // TODO: Calculate based on node count and types
    };
  }

  /**
   * Get all start nodes in the diagram
   */
  getStartNodes(): string[] {
    return Object.entries(this.context.nodesById)
      .filter(([_, node]) => node.type === 'start')
      .map(([nodeId, _]) => nodeId);
  }

  /**
   * Get the next nodes to execute based on current node output
   */
  getNextNodes(nodeId: string, output?: unknown): string[] {
    const nextNodes: string[] = [];
    const outgoing = this.context.outgoingArrows[nodeId] || [];

    for (const arrow of outgoing) {
      const typedArrow = arrow as DiagramArrow;
      const targetId = this.getArrowTarget(typedArrow);
      if (!targetId) continue;

      // Handle conditional arrows
      if (this.isConditionalArrow(typedArrow) && typeof output === 'boolean') {
        const arrowLabel = this.getArrowLabel(typedArrow);
        if ((arrowLabel === 'true' && output) || (arrowLabel === 'false' && !output)) {
          nextNodes.push(targetId);
        }
      } else {
        // Non-conditional arrow
        nextNodes.push(targetId);
      }
    }

    return nextNodes;
  }

  /**
   * Identify nodes that can be executed in parallel
   */
  detectParallelExecutionOpportunities(): Set<string>[] {
    const parallelGroups: Set<string>[] = [];
    const processed = new Set<string>();

    while (processed.size < Object.keys(this.context.nodesById).length) {
      // Find nodes with no unprocessed dependencies
      const readyNodes = new Set<string>();

      for (const nodeId of Object.keys(this.context.nodesById)) {
        if (processed.has(nodeId)) continue;

        // Check if all dependencies are processed
        const deps = new Set(this.dependencyResolver.getDependencies(nodeId));
        if (this.isSubset(deps, processed)) {
          readyNodes.add(nodeId);
        }
      }

      if (readyNodes.size === 0) {
        // Handle cycles or disconnected nodes
        const remaining = new Set(Object.keys(this.context.nodesById));
        for (const processed_node of processed) {
          remaining.delete(processed_node);
        }
        if (remaining.size > 0) {
          const firstRemaining = remaining.values().next().value;
          if (firstRemaining) {
            readyNodes.add(firstRemaining);
          }
        }
      }

      if (readyNodes.size > 0) {
        parallelGroups.push(readyNodes);
        for (const node of readyNodes) {
          processed.add(node);
        }
      }
    }

    return parallelGroups;
  }

  /**
   * Validate if an execution path is valid
   */
  validateExecutionPath(path: string[]): { is_valid: boolean; errors: string[]; warnings: string[] } {
    const validationResult = {
      is_valid: true,
      errors: [] as string[],
      warnings: [] as string[]
    };

    const executed = new Set<string>();

    for (let i = 0; i < path.length; i++) {
      const nodeId = path[i];
      
      if (!nodeId || !(nodeId in this.context.nodesById)) {
        validationResult.is_valid = false;
        validationResult.errors.push(`Node ${nodeId} not found in diagram`);
        continue;
      }

      // Check dependencies
      if (nodeId) {
        const deps = new Set(this.dependencyResolver.getDependencies(nodeId));
        const missingDeps = new Set([...deps].filter(dep => !executed.has(dep)));

        if (missingDeps.size > 0) {
          validationResult.is_valid = false;
          validationResult.errors.push(
            `Node ${nodeId} at position ${i} has unmet dependencies: ${Array.from(missingDeps).join(', ')}`
          );
        }

        executed.add(nodeId);
      }
    }

    return validationResult;
  }

  /**
   * Determine optimal execution order using topological sort
   */
  determineExecutionOrder(): string[] {
    return this.dependencyResolver.topologicalSort();
  }

  // Private helper methods

  private isConditionalArrow(arrow: DiagramArrow): boolean {
    const label = this.getArrowLabel(arrow);
    return label === 'true' || label === 'false';
  }

  private getArrowLabel(arrow: DiagramArrow): string {
    // Check multiple possible locations for the label
    let label = arrow.label || '';
    if (!label) {
      // Check sourceHandle for condition output patterns
      const sourceHandle = arrow.sourceHandle || '';
      if (sourceHandle.endsWith('-output-true')) {
        label = 'true';
      } else if (sourceHandle.endsWith('-output-false')) {
        label = 'false';
      } else {
        // Check arrow data
        label = arrow.data?.label || '';
      }
    }

    return typeof label === 'string' ? label.toLowerCase() : '';
  }

  private getAllArrows(): DiagramArrow[] {
    const allArrows: DiagramArrow[] = [];
    for (const arrows of Object.values(this.context.outgoingArrows)) {
      allArrows.push(...(arrows as DiagramArrow[]));
    }
    return allArrows;
  }

  private getArrowTarget(arrow: DiagramArrow): string | null {
    return arrow.target || null;
  }

  private isSubset<T>(subset: Set<T>, superset: Set<T>): boolean {
    for (const elem of subset) {
      if (!superset.has(elem)) {
        return false;
      }
    }
    return true;
  }
}